File: custom-domain-dns-verification-failure/node/custom-domain-dns-check.js
/**
 * Detect a GitHub Pages custom domain DNS check failure and repair it via Cloudflare.
 * Safe to run on a schedule. Stays in dry run until DRY_RUN=false.
 */
import { pathToFileURL } from "node:url";

export const GITHUB_PAGES_A_RECORDS = new Set([
  "185.199.108.153",
  "185.199.109.153",
  "185.199.110.153",
  "185.199.111.153",
]);

export function diagnosePagesDns(apexARecords, wwwCnameTarget, requiredARecords, expectedCnameSuffix) {
  // Pure decision function. No network, no I/O.
  const missing = new Set([...requiredARecords].filter((ip) => !apexARecords.has(ip)));
  const extra = new Set([...apexARecords].filter((ip) => !requiredARecords.has(ip)));
  const apexOk = missing.size === 0 && extra.size === 0;
  const suffix = expectedCnameSuffix.replace(/^\./, "");
  const wwwOk = Boolean(wwwCnameTarget) && wwwCnameTarget.replace(/\.$/, "").endsWith(suffix);
  return {
    apex_ok: apexOk,
    apex_missing: missing,
    apex_extra: extra,
    www_ok: wwwOk,
    www_target: wwwCnameTarget,
  };
}

export async function run() {
  // Imported lazily so the pure function above can be tested with no
  // network modules touched at all.
  const dns = await import("node:dns");
  const resolvePromises = dns.promises;

  const domain = process.env.DNS_DOMAIN || "example.com";
  const githubHostname = process.env.GITHUB_PAGES_HOSTNAME || "yourusername.github.io";
  const dryRun = (process.env.DRY_RUN || "true").toLowerCase() === "true";

  const zoneId = process.env.CLOUDFLARE_ZONE_ID;
  const apiToken = process.env.CLOUDFLARE_API_TOKEN;

  let apexARecords = new Set();
  try {
    apexARecords = new Set(await resolvePromises.resolve4(domain));
  } catch (err) {
    if (err.code !== "ENOTFOUND" && err.code !== "ENODATA") throw err;
  }

  const wwwName = `www.${domain}`;
  let wwwCnameTarget = null;
  try {
    const cnames = await resolvePromises.resolveCname(wwwName);
    wwwCnameTarget = cnames[0] || null;
  } catch (err) {
    if (err.code !== "ENOTFOUND" && err.code !== "ENODATA") throw err;
  }

  const report = diagnosePagesDns(apexARecords, wwwCnameTarget, GITHUB_PAGES_A_RECORDS, ".github.io");
  console.log(`Apex ok=${report.apex_ok} missing=${[...report.apex_missing]} extra=${[...report.apex_extra]}`);
  console.log(`www ok=${report.www_ok} target=${report.www_target}`);

  if (report.apex_ok && report.www_ok) {
    console.log("Nothing to repair. DNS matches GitHub Pages requirements.");
    return;
  }

  const headers = { Authorization: `Bearer ${apiToken}`, "Content-Type": "application/json" };
  const base = `https://api.cloudflare.com/client/v4/zones/${zoneId}/dns_records`;

  if (!report.apex_ok) {
    console.log(`Apex ${domain} is wrong. ${dryRun ? "Would" : "Will"} replace the A records with the four GitHub Pages IPs.`);
    if (!dryRun) {
      const existing = await (await fetch(`${base}?type=A&name=${domain}`, { headers })).json();
      for (const record of existing.result || []) {
        await fetch(`${base}/${record.id}`, { method: "DELETE", headers });
      }
      for (const ip of [...GITHUB_PAGES_A_RECORDS].sort()) {
        await fetch(base, { method: "POST", headers, body: JSON.stringify({ type: "A", name: domain, content: ip, ttl: 300 }) });
      }
    }
  }

  if (!report.www_ok) {
    console.log(`www.${domain} is wrong. ${dryRun ? "Would" : "Will"} replace it with a CNAME to ${githubHostname}.`);
    if (!dryRun) {
      const existing = await (await fetch(`${base}?name=${wwwName}`, { headers })).json();
      for (const record of existing.result || []) {
        await fetch(`${base}/${record.id}`, { method: "DELETE", headers });
      }
      await fetch(base, { method: "POST", headers, body: JSON.stringify({ type: "CNAME", name: wwwName, content: githubHostname, ttl: 300 }) });
    }
  }

  console.log("Done.");
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: custom-domain-dns-verification-failure/python/custom_domain_dns_check.py
"""Detect a GitHub Pages custom domain DNS check failure and repair it via Cloudflare.
Safe to run on a schedule. Stays in dry run until DRY_RUN=false.
"""
import os
import logging

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("custom_domain_dns_check")

GITHUB_PAGES_A_RECORDS = {
    "185.199.108.153",
    "185.199.109.153",
    "185.199.110.153",
    "185.199.111.153",
}


def diagnose_pages_dns(apex_a_records: set, www_cname_target, required_a_records: set, expected_cname_suffix: str) -> dict:
    """Pure decision function. No network, no I/O."""
    missing = required_a_records - apex_a_records
    extra = apex_a_records - required_a_records
    apex_ok = not missing and not extra
    www_ok = bool(www_cname_target) and www_cname_target.rstrip(".").endswith(expected_cname_suffix.lstrip("."))
    return {
        "apex_ok": apex_ok,
        "apex_missing": missing,
        "apex_extra": extra,
        "www_ok": www_ok,
        "www_target": www_cname_target,
    }


def run():
    # Imported lazily so the pure function above can be tested with no
    # network libraries installed at all.
    import dns.resolver
    import requests

    domain = os.environ["DNS_DOMAIN"]
    github_hostname = os.environ.get("GITHUB_PAGES_HOSTNAME", "yourusername.github.io")
    dry_run = os.environ.get("DRY_RUN", "true").lower() == "true"

    zone_id = os.environ["CLOUDFLARE_ZONE_ID"]
    api_token = os.environ["CLOUDFLARE_API_TOKEN"]

    try:
        apex_answer = dns.resolver.resolve(domain, "A")
        apex_a_records = {str(r) for r in apex_answer}
    except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer):
        apex_a_records = set()

    www_name = f"www.{domain}"
    try:
        cname_answer = dns.resolver.resolve(www_name, "CNAME")
        www_cname_target = str(cname_answer[0].target).rstrip(".")
    except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer):
        www_cname_target = None

    report = diagnose_pages_dns(apex_a_records, www_cname_target, GITHUB_PAGES_A_RECORDS, ".github.io")
    log.info("Apex ok=%s missing=%s extra=%s", report["apex_ok"], report["apex_missing"], report["apex_extra"])
    log.info("www ok=%s target=%s", report["www_ok"], report["www_target"])

    if report["apex_ok"] and report["www_ok"]:
        log.info("Nothing to repair. DNS matches GitHub Pages requirements.")
        return

    headers = {"Authorization": f"Bearer {api_token}", "Content-Type": "application/json"}
    base = f"https://api.cloudflare.com/client/v4/zones/{zone_id}/dns_records"

    if not report["apex_ok"]:
        log.info("Apex %s is wrong. %s replace the A records with the four GitHub Pages IPs.",
                  domain, "Would" if dry_run else "Will")
        if not dry_run:
            existing = requests.get(base, headers=headers, params={"type": "A", "name": domain}, timeout=30).json()
            for record in existing.get("result", []):
                requests.delete(f"{base}/{record['id']}", headers=headers, timeout=30).raise_for_status()
            for ip in sorted(GITHUB_PAGES_A_RECORDS):
                requests.post(base, headers=headers, json={"type": "A", "name": domain, "content": ip, "ttl": 300}, timeout=30).raise_for_status()

    if not report["www_ok"]:
        log.info("www.%s is wrong. %s replace it with a CNAME to %s.",
                  domain, "Would" if dry_run else "Will", github_hostname)
        if not dry_run:
            existing = requests.get(base, headers=headers, params={"name": www_name}, timeout=30).json()
            for record in existing.get("result", []):
                requests.delete(f"{base}/{record['id']}", headers=headers, timeout=30).raise_for_status()
            requests.post(base, headers=headers, json={"type": "CNAME", "name": www_name, "content": github_hostname, "ttl": 300}, timeout=30).raise_for_status()

    log.info("Done.")


if __name__ == "__main__":
    run()
